File: concepts/04-queues/python/queue.py
"""
Queue Implementation in Python
"""

from collections import deque

class Queue:
    """Array-based Queue Implementation"""
    
    def __init__(self):
        self.items = []
    
    def enqueue(self, item):
        """Add element to rear - O(1)"""
        self.items.append(item)
    
    def dequeue(self):
        """Remove and return front element - O(1)"""
        if self.is_empty():
            raise IndexError("Queue is empty")
        return self.items.pop(0)
    
    def front(self):
        """Return front element without removing - O(1)"""
        if self.is_empty():
            raise IndexError("Queue is empty")
        return self.items[0]
    
    def is_empty(self):
        """Check if queue is empty - O(1)"""
        return len(self.items) == 0
    
    def size(self):
        """Return size of queue - O(1)"""
        return len(self.items)
    
    def display(self):
        """Display all elements"""
        return self.items.copy()


class QueueNode:
    """Node for linked list-based queue"""
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedQueue:
    """Linked List-based Queue Implementation"""
    
    def __init__(self):
        self.front_node = None
        self.rear_node = None
        self.size = 0
    
    def enqueue(self, item):
        """Add element to rear - O(1)"""
        new_node = QueueNode(item)
        if self.is_empty():
            self.front_node = self.rear_node = new_node
        else:
            self.rear_node.next = new_node
            self.rear_node = new_node
        self.size += 1
    
    def dequeue(self):
        """Remove and return front element - O(1)"""
        if self.is_empty():
            raise IndexError("Queue is empty")
        data = self.front_node.data
        self.front_node = self.front_node.next
        if self.front_node is None:
            self.rear_node = None
        self.size -= 1
        return data
    
    def front(self):
        """Return front element without removing - O(1)"""
        if self.is_empty():
            raise IndexError("Queue is empty")
        return self.front_node.data
    
    def is_empty(self):
        """Check if queue is empty - O(1)"""
        return self.front_node is None
    
    def size(self):
        """Return size of queue - O(1)"""
        return self.size


class CircularQueue:
    """Circular Queue Implementation"""
    
    def __init__(self, capacity):
        self.capacity = capacity
        self.items = [None] * capacity
        self.front = 0
        self.rear = -1
        self.size = 0
    
    def enqueue(self, item):
        """Add element - O(1)"""
        if self.is_full():
            raise OverflowError("Queue is full")
        self.rear = (self.rear + 1) % self.capacity
        self.items[self.rear] = item
        self.size += 1
    
    def dequeue(self):
        """Remove and return element - O(1)"""
        if self.is_empty():
            raise IndexError("Queue is empty")
        item = self.items[self.front]
        self.items[self.front] = None
        self.front = (self.front + 1) % self.capacity
        self.size -= 1
        return item
    
    def front(self):
        """Return front element - O(1)"""
        if self.is_empty():
            raise IndexError("Queue is empty")
        return self.items[self.front]
    
    def is_empty(self):
        """Check if empty - O(1)"""
        return self.size == 0
    
    def is_full(self):
        """Check if full - O(1)"""
        return self.size == self.capacity


class PriorityQueue:
    """Priority Queue using list (min-heap simulation)"""
    
    def __init__(self):
        self.items = []
    
    def enqueue(self, item, priority):
        """Add element with priority - O(n)"""
        self.items.append((priority, item))
        self.items.sort(key=lambda x: x[0])
    
    def dequeue(self):
        """Remove highest priority element - O(1)"""
        if self.is_empty():
            raise IndexError("Queue is empty")
        return self.items.pop(0)[1]
    
    def is_empty(self):
        """Check if empty - O(1)"""
        return len(self.items) == 0


class Deque:
    """Double-ended Queue"""
    
    def __init__(self):
        self.items = []
    
    def add_front(self, item):
        """Add to front - O(n)"""
        self.items.insert(0, item)
    
    def add_rear(self, item):
        """Add to rear - O(1)"""
        self.items.append(item)
    
    def remove_front(self):
        """Remove from front - O(n)"""
        if self.is_empty():
            raise IndexError("Deque is empty")
        return self.items.pop(0)
    
    def remove_rear(self):
        """Remove from rear - O(1)"""
        if self.is_empty():
            raise IndexError("Deque is empty")
        return self.items.pop()
    
    def is_empty(self):
        """Check if empty - O(1)"""
        return len(self.items) == 0
    
    def size(self):
        """Return size - O(1)"""
        return len(self.items)


# Example usage
if __name__ == "__main__":
    # Simple Queue
    print("--- Simple Queue ---")
    queue = Queue()
    queue.enqueue(1)
    queue.enqueue(2)
    queue.enqueue(3)
    print("Queue:", queue.display())
    print("Front:", queue.front())
    print("Dequeue:", queue.dequeue())
    print("After dequeue:", queue.display())
    
    # Linked Queue
    print("\n--- Linked Queue ---")
    linked_queue = LinkedQueue()
    linked_queue.enqueue(10)
    linked_queue.enqueue(20)
    linked_queue.enqueue(30)
    print("Size:", linked_queue.size())
    print("Front:", linked_queue.front())
    print("Dequeue:", linked_queue.dequeue())
    
    # Circular Queue
    print("\n--- Circular Queue ---")
    circular_queue = CircularQueue(5)
    for i in range(5):
        circular_queue.enqueue(i)
    print("Front:", circular_queue.front())
    print("Dequeue:", circular_queue.dequeue())
    circular_queue.enqueue(99)
    
    # Priority Queue
    print("\n--- Priority Queue ---")
    pq = PriorityQueue()
    pq.enqueue("Task 1", 3)
    pq.enqueue("Task 2", 1)
    pq.enqueue("Task 3", 2)
    print("Dequeue:", pq.dequeue())  # Task 2 (priority 1)
    
    # Deque
    print("\n--- Deque ---")
    dq = Deque()
    dq.add_rear(1)
    dq.add_rear(2)
    dq.add_front(0)
    print("Remove front:", dq.remove_front())
    print("Remove rear:", dq.remove_rear())

File: concepts/04-queues/node/queue.js
/**
 * Queue Implementation in Node.js
 */

class Queue {
    /**
     * Array-based Queue Implementation
     */
    constructor() {
        this.items = [];
    }

    /**
     * Add element to rear - O(1)
     */
    enqueue(item) {
        this.items.push(item);
    }

    /**
     * Remove and return front element - O(1)
     */
    dequeue() {
        if (this.isEmpty()) {
            throw new Error("Queue is empty");
        }
        return this.items.shift();
    }

    /**
     * Return front element without removing - O(1)
     */
    front() {
        if (this.isEmpty()) {
            throw new Error("Queue is empty");
        }
        return this.items[0];
    }

    /**
     * Check if queue is empty - O(1)
     */
    isEmpty() {
        return this.items.length === 0;
    }

    /**
     * Return size of queue - O(1)
     */
    size() {
        return this.items.length;
    }

    /**
     * Display all elements
     */
    display() {
        return [...this.items];
    }
}

class QueueNode {
    /**
     * Node for linked list-based queue
     */
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedQueue {
    /**
     * Linked List-based Queue Implementation
     */
    constructor() {
        this.frontNode = null;
        this.rearNode = null;
        this.size = 0;
    }

    /**
     * Add element to rear - O(1)
     */
    enqueue(item) {
        const newNode = new QueueNode(item);
        if (this.isEmpty()) {
            this.frontNode = this.rearNode = newNode;
        } else {
            this.rearNode.next = newNode;
            this.rearNode = newNode;
        }
        this.size++;
    }

    /**
     * Remove and return front element - O(1)
     */
    dequeue() {
        if (this.isEmpty()) {
            throw new Error("Queue is empty");
        }
        const data = this.frontNode.data;
        this.frontNode = this.frontNode.next;
        if (this.frontNode === null) {
            this.rearNode = null;
        }
        this.size--;
        return data;
    }

    /**
     * Return front element without removing - O(1)
     */
    front() {
        if (this.isEmpty()) {
            throw new Error("Queue is empty");
        }
        return this.frontNode.data;
    }

    /**
     * Check if queue is empty - O(1)
     */
    isEmpty() {
        return this.frontNode === null;
    }

    /**
     * Return size of queue - O(1)
     */
    size() {
        return this.size;
    }
}

class CircularQueue {
    /**
     * Circular Queue Implementation
     */
    constructor(capacity) {
        this.capacity = capacity;
        this.items = new Array(capacity).fill(null);
        this.front = 0;
        this.rear = -1;
        this.size = 0;
    }

    /**
     * Add element - O(1)
     */
    enqueue(item) {
        if (this.isFull()) {
            throw new Error("Queue is full");
        }
        this.rear = (this.rear + 1) % this.capacity;
        this.items[this.rear] = item;
        this.size++;
    }

    /**
     * Remove and return element - O(1)
     */
    dequeue() {
        if (this.isEmpty()) {
            throw new Error("Queue is empty");
        }
        const item = this.items[this.front];
        this.items[this.front] = null;
        this.front = (this.front + 1) % this.capacity;
        this.size--;
        return item;
    }

    /**
     * Return front element - O(1)
     */
    front() {
        if (this.isEmpty()) {
            throw new Error("Queue is empty");
        }
        return this.items[this.front];
    }

    /**
     * Check if empty - O(1)
     */
    isEmpty() {
        return this.size === 0;
    }

    /**
     * Check if full - O(1)
     */
    isFull() {
        return this.size === this.capacity;
    }
}

class PriorityQueue {
    /**
     * Priority Queue using array (min-heap simulation)
     */
    constructor() {
        this.items = [];
    }

    /**
     * Add element with priority - O(n)
     */
    enqueue(item, priority) {
        this.items.push({ item, priority });
        this.items.sort((a, b) => a.priority - b.priority);
    }

    /**
     * Remove highest priority element - O(1)
     */
    dequeue() {
        if (this.isEmpty()) {
            throw new Error("Queue is empty");
        }
        return this.items.shift().item;
    }

    /**
     * Check if empty - O(1)
     */
    isEmpty() {
        return this.items.length === 0;
    }
}

class Deque {
    /**
     * Double-ended Queue
     */
    constructor() {
        this.items = [];
    }

    /**
     * Add to front - O(n)
     */
    addFront(item) {
        this.items.unshift(item);
    }

    /**
     * Add to rear - O(1)
     */
    addRear(item) {
        this.items.push(item);
    }

    /**
     * Remove from front - O(n)
     */
    removeFront() {
        if (this.isEmpty()) {
            throw new Error("Deque is empty");
        }
        return this.items.shift();
    }

    /**
     * Remove from rear - O(1)
     */
    removeRear() {
        if (this.isEmpty()) {
            throw new Error("Deque is empty");
        }
        return this.items.pop();
    }

    /**
     * Check if empty - O(1)
     */
    isEmpty() {
        return this.items.length === 0;
    }

    /**
     * Return size - O(1)
     */
    size() {
        return this.items.length;
    }
}

// Example usage
if (require.main === module) {
    // Simple Queue
    console.log("--- Simple Queue ---");
    const queue = new Queue();
    queue.enqueue(1);
    queue.enqueue(2);
    queue.enqueue(3);
    console.log("Queue:", queue.display());
    console.log("Front:", queue.front());
    console.log("Dequeue:", queue.dequeue());
    console.log("After dequeue:", queue.display());

    // Linked Queue
    console.log("\n--- Linked Queue ---");
    const linkedQueue = new LinkedQueue();
    linkedQueue.enqueue(10);
    linkedQueue.enqueue(20);
    linkedQueue.enqueue(30);
    console.log("Size:", linkedQueue.size());
    console.log("Front:", linkedQueue.front());
    console.log("Dequeue:", linkedQueue.dequeue());

    // Circular Queue
    console.log("\n--- Circular Queue ---");
    const circularQueue = new CircularQueue(5);
    for (let i = 0; i < 5; i++) {
        circularQueue.enqueue(i);
    }
    console.log("Front:", circularQueue.front());
    console.log("Dequeue:", circularQueue.dequeue());
    circularQueue.enqueue(99);

    // Priority Queue
    console.log("\n--- Priority Queue ---");
    const pq = new PriorityQueue();
    pq.enqueue("Task 1", 3);
    pq.enqueue("Task 2", 1);
    pq.enqueue("Task 3", 2);
    console.log("Dequeue:", pq.dequeue()); // Task 2 (priority 1)

    // Deque
    console.log("\n--- Deque ---");
    const dq = new Deque();
    dq.addRear(1);
    dq.addRear(2);
    dq.addFront(0);
    console.log("Remove front:", dq.removeFront());
    console.log("Remove rear:", dq.removeRear());
}

module.exports = { Queue, LinkedQueue, CircularQueue, PriorityQueue, Deque };

